File: Day011_2025/Q011_GFG.py
class Solution:
    def longestUniqueSubstr(self, s: str) -> int:
        # Code here
        # Q. Longest substring with distinct characters
        char_map = {}
        max_length = 0
        j = 0

        for i in range(len(s)):
            char_map[s[i]] = char_map.get(s[i], 0) + 1
            while char_map[s[i]] > 1:
                char_map[s[j]] -= 1
                if char_map[s[j]] == 0:
                    del char_map[s[j]]
                j += 1
            max_length = max(max_length, i - j + 1)

        return max_length


# Example usage
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        s = input()
        sol = Solution()
        print(sol.longestUniqueSubstr(s))

File: Day011_2025/Q011_GFG.java
import java.util.*;

class Q011_GFG {
    public int longestUniqueSubstr(String s) {
        // Code here
        // Q. Longest substring with distinct characters
        Map<Character, Integer> charMap = new HashMap<>();
        int maxLength = 0;
        int j = 0;

        for (int i = 0; i < s.length(); i++) {
            charMap.put(s.charAt(i), charMap.getOrDefault(s.charAt(i), 0) + 1);
            while (charMap.get(s.charAt(i)) > 1) {
                char startChar = s.charAt(j);
                charMap.put(startChar, charMap.get(startChar) - 1);
                if (charMap.get(startChar) == 0) {
                    charMap.remove(startChar);
                }
                j++;
            }
            maxLength = Math.max(maxLength, i - j + 1);
        }
        return maxLength;
    }

    public static void main(String[] args) {
        Q011_GFG sol = new Q011_GFG();
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            String s = sc.next();
            System.out.println(sol.longestUniqueSubstr(s));
        }
        sc.close();
    }
}
